### ISSUE TITLE
Sprites not being redrawn correctly when style's sprite sheet has been updated

### ISSUE THREAD

**Issue by friedbunny** [mapbox/mapbox-gl-native#4117]
The `mapbox-streets-v7` source has just gone live [in all Mapbox machine styles](https://github.com/mapbox/mapbox-gl-styles/issues/233) and now all styles have [complete sprite sets](https://github.com/mapbox/mapbox-gl-styles/issues/241).

These sprite changes have exposed a bug: after loading a new sprite sheet, we don't seem to update the coordinates of the individual icons correctly. Reloading the style fixes the misdrawn icons.

![img_4047](https://cloud.githubusercontent.com/assets/1198851/13346739/5645d7d8-dc36-11e5-9682-e9d4e38ece05.PNG)

/cc @ansis @kkaefer

**Comment by friedbunny** [comment]
This still happens and can be amusing:

![simulator screen shot may 21 2016 2 26 27 pm](https://cloud.githubusercontent.com/assets/1198851/15482193/48982f0e-20fc-11e6-9c2e-b2d7f34f3929.png)

**Comment by jfirebaugh** [comment]
Speculation is this happens if you get a stale `sprite.png` and fresh `sprite.json`, or vice versa. https://github.com/mapbox/mapbox-gl-style-spec/issues/220 is the solution.

**Comment by dgohlke** [comment]
We're seeing the sprite.png getting out of sync against the sprite.json as well. We've only noticed this in our iOS app which is using mapbox sdk version 3.7.3.

We expected to see this border around the 5:
![image](https://user-images.githubusercontent.com/9354850/36328926-025e1b46-1332-11e8-9f08-4ce4d3c980b7.png)

But we instead saw this border around the 1 (ignore the numerical & underlying image differences):
![image](https://user-images.githubusercontent.com/9354850/36328938-12451fa0-1332-11e8-8047-2534cd293fa0.png)

Compare the following two sprites (before on top, after on bottom), you'll see that where the ![image](https://user-images.githubusercontent.com/9354850/36328958-276ee794-1332-11e8-8912-20fdcc799ce5.png) was, now there's ![image](https://user-images.githubusercontent.com/9354850/36328973-310b9c7a-1332-11e8-981d-868dc8a3989d.png). So that tells me that the image was updated but the SDK was still using the old sprite mapping json.

Before change:
![image](https://user-images.githubusercontent.com/9354850/36328978-34b37ef6-1332-11e8-9563-159b26680807.png)

After change:
![image](https://user-images.githubusercontent.com/9354850/36328987-3d093942-1332-11e8-95e2-5c835970bf28.png)

You can see that the sprite slice is being loaded at the same x,y coordinates, but the sprite.png was changed. This means the sprite.json is stale while the sprite.png is fresh.

This happens after adding a SVG to the sprite via mapbox studio and publishing the style. The only fix we've found is to uninstall and reinstall the app. This bug has earned us at least one 1-star review, so its one we'd like to address immediately.

------------------------
@ryanbaumann Thank you for taking my email after chatting at the Bikeshare hack night in DC.

**Comment by lucaswoj** [comment]
https://github.com/mapbox/mapbox-gl-js/issues/4086 is a long term solution that'll make it less likely for bugs like this in the future but it isn't necessary or sufficient to fix this particular bug. Given that this is affecting customers, should we look into a short term fix?

**Comment by lucaswoj** [comment]
Clarification by @jfirebaugh via Slack:

> My thinking was more that our caching strategy inherently permits this to happen without any intermediate caches necessarily being "stale" as defined by the headers we use. But a bug that produces incorrectly stale results out of the cache is also a possibility.
>
> As an example of the latter possibility: https://github.com/mapbox/mapbox-gl-native/issues/10015
>
> Maybe it's even possible that this is another manifestation of the same issue.
>
> Ah, probably not, that was fixed and backported to Android SDK v4.2.0 and iOS SDK v3.7.0.

**Comment by lucaswoj** [comment]
One potential workaround is to use the `fresh=true` parameter on the style URL. I'm not exactly sure how GL native handles this setting but there's a chance it cuts out the caching behaviour that's causing the bug.

**Comment by christinedumont** [comment]
Little update. We are running Android 5.5.0 and iOS 3.7.3 and we are still experiencing this issue. 
On iOS, everytime the app comes to foreground, we set the style URL with a query param that contains the current timestamp, so it should be "forcing" a new style URL. It didn't fix it. We will try the fresh=true.

`                
if let url = URL.init(string: "\(BuildTargetConfigLoader.buildTargetConfig.mapStyleURL.absoluteString)?cacheBreak=\(Date().timeIntervalSince1970)") {
                    self?.styleURL = url
}
`

**Comment by chloekraw** [comment]
This is still an issue

**Comment by ansis** [comment]
I think mapbox/mapbox-gl-js#4086 would be the best way to fix this but

Another approach would be to store every version of a sprite and have the api be able to serve specific versions. Every time you modify the sprite it would increment the version (or take a hash).
```
"sprite": "mapbox://sprites/mapbox/bright-v8?version=123"
```

**Comment by chloekraw** [comment]
👋 We've deployed a server-side change in early April that we believe should resolve or significantly reduce instances of this issue. If you're still seeing instances of sprite json/pngs going out of sync in production, could you post back to this thread to let us know?

@dgohlke @christinedumont

**Comment by tsemerad** [comment]
I just ran into this issue in production today. I'm using the iOS SDK v4.3.0, if that matters. This is really disruptive to our users.

I see that there is now an ability to clear the "ambient cache" #4376. I think this might give me the ability to clear out the cached sprite sheet, but is there a less heavy-handed approach? I really don't want to have to clear the entire offline tile cache. I'd like to just get the latest sprite .json and .png whenever I launch the map.

For the moment, I feel like I can't update my map's sprite sheet without risking interfering with my current users' operation of the app.

**Comment by asheemmamoowala** [comment]
@tsemerad Thank you for reporting your issue here. Could you confirm whether you updated your spritesheet in Mapbox Studio? If your style is public, could you share a link to it as well?

**Comment by tsemerad** [comment]
@asheemmamoowala Yes I updated the spritesheet in Mapbox Studio. The style is private unfortunately. The issue is sporadic (only happened on 1 of our test devices) but restarting the app doesn't fix it - only reinstalling the app.

**Comment by kkaefer** [comment]
@tsemerad could you please reach out to [Mapbox suppport](https://support.mapbox.com/hc/en-us/requests/new?ticket_form_id=360000291231) with your style URL and mention this issue?

**Comment by tsemerad** [comment]
@kkaefer I sent a support request at the end of last week. Thanks for the suggestion. I'll post here again when or if anything comes of it.
